core/src/main/java/org/elasticsearch/index/analysis/SynonymTokenFilterFactory.java

```diff
@@ -23,82 +23,151 @@
+            }
+
         } else {
             throw new IllegalArgumentException("synonym requires either `synonyms` or `synonyms_path` to be configured");
         }
 
         this.ignoreCase = settings.getAsBoolean("ignore_case", false);
-        boolean expand = settings.getAsBoolean("expand", true);
+        this.expand = settings.getAsBoolean("expand", true);
 
-        String tokenizerName = settings.get("tokenizer", "whitespace");
-        final TokenizerFactory tokenizerFactory = tokenizerFactories.get(tokenizerName);
-        if (tokenizerFactory == null) {
-            throw new IllegalArgumentException("failed to find tokenizer [" + tokenizerName + "] for synonym token filter");
-        }
+        // for backward compatibility
+        String tokenizerName = settings.get("tokenizer", "");
 
-        Analyzer analyzer = new Analyzer() {
-            @Override
-            protected TokenStreamComponents createComponents(String fieldName) {
-                Tokenizer tokenizer = tokenizerFactory == null ? new WhitespaceTokenizer() : tokenizerFactory.create();
-                TokenStream stream = ignoreCase ? new LowerCaseFilter(tokenizer) : tokenizer;
-                return new TokenStreamComponents(tokenizer, stream);
+        if (tokenizerName.length() > 0) {
+            TokenizerFactory tokenizerFactory = tokenizerFactories.get(tokenizerName);
+            if (tokenizerFactory == null) {
+                AnalysisModule.AnalysisProvider<TokenizerFactory> tokenizerFactoryFactory =
+                    analysisRegistry.getTokenizerProvider(tokenizerName);
+                tokenizerFactory = tokenizerFactoryFactory.get(env, tokenizerName);
             }
-        };
-
-        try {
-            SynonymMap.Builder parser = null;
-
-            if ("wordnet".equalsIgnoreCase(settings.get("format"))) {
-                parser = new WordnetSynonymParser(true, expand, analyzer);
-                ((WordnetSynonymParser) parser).parse(rulesReader);
-            } else {
-                parser = new SolrSynonymParser(true, expand, analyzer);
-                ((SolrSynonymParser) parser).parse(rulesReader);
+            if (tokenizerFactory == null) {
+                throw new IllegalArgumentException("failed to find tokenizer [" + tokenizerName + "] for synonym token filter");
             }
-
-            synonymMap = parser.build();
-        } catch (Exception e) {
-            throw new IllegalArgumentException("failed to build synonyms", e);
+            this.tokenizerFactory = tokenizerFactory;
+        } else {
+            this.tokenizerFactory = null;
         }
+
+        this.format = settings.get("format", "");
     }
 
     @Override
     public TokenStream create(TokenStream tokenStream) {
         // fst is null means no synonyms
         return synonymMap.fst == null ? tokenStream : new SynonymFilter(tokenStream, synonymMap, ignoreCase);
     }
+
+    private final String format;
+    private final boolean expand;
```

Comment:
Can you move these class variable definitions up to the top of the class? It's weird to see them after function definitions